File: python/101. Symmetric Tree.py
#coding=utf-8
__author__ = 'xuxuan'
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return True if root is None or self.f(root.left,root.right) else False
    def f(self,left,right):
        if left is None or right is None:
            return left is right
        return left.val==right.val and self.f(left.left,right.right) and self.f(left.right,right.left)

File: ruby/101. Symmetric Tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def is_symmetric(root)
    return ! root || is_same(root.left,root.right)
end

def is_same(left,right)
    return left && right ? left.val==right.val && is_same(left.left,right.right) && is_same(left.right,right.left) : left==right
end